### ISSUE TITLE
Feature Request: Easier to extend base RNN implementation

### ISSUE THREAD

**Issue by csarofeen** [pytorch/pytorch#711]
Currently base RNN class/functions are hard to extend. If someone would like to extend LSTM with new features to pytorch they would have to modify:
AutogradRNN (nn/_functions/rnn.py)
StackedRNN (nn/_functions/rnn.py)
RNNBase (nn/modules/rnn.py)
Furthermore, the default RNN implementation is restrictive, enforcing every stacked RNN layer to be exactly the same. 

I was thinking it may be worthwhile to instead have an RNN driver abstract RNN base class.

RNN Driver would be similar in function to recurrent, stackedRNN, and AutogradRNN functions but would be dependent on abstract RNN class but independent of specific RNN definitions.

The code would look something like...

```
class RNNDriver(nn.Module):        
    #constructor could either take an RNN or list of RNN layers
    def __init__(self, inputRNN, num_layers=1):
        if not isinstance(inputRNN, list):
            self.rnns = [inputRNN]
            for i in range(num_layers-1):
                self.rnns.append(inputRNN.clone())
        else:
            assert len(inputRNN) == num_layers, "RNN list length must be equal to num_layers"
        
    #Parameters call to group parameters of all layers
    def parameters(self):
        memo = set()
        for rnn in self.rnns:
            for p in rnn.parameters(memo):
                yield p
                
    def forward(self, input, train=True, batch_first=False, dropout=0, bidirectional=False):
        ...

    def initHidden(self, bsz):
        for rnn in self.rnns:
            rnn.initHidden(bsz)

    def resetHidden(self, bsz):
        for rnn in self.rnns:
            rnn.resetHidden(bsz)

    def initInference(self, bsz):    
        for rnn in self.rnns:
            rnn.initInference(bsz)

class RNNBase(nn.Module):
#Base initialization could be for a simple RNN layer or could be empty
    def __init__(self, input_size, hidden_size):
        super(RNNBase, self).__init__()
        self.hidden_size = hidden_size
        self.input_size = input_size

        self.w_ih = nn.Parameter(torch.Tensor(hidden_size, input_size))
        self.b_ih = nn.Parameter(torch.Tensor(hidden_size))
        self.w_hh = nn.Parameter(torch.Tensor(hidden_size, hidden_size))

        self.hidden = None
        self.reset_parameters()

    def reset_parameters(self, feature_size):
            stdv = 1.0 / math.sqrt(feature_size)
            for weight in self.parameters():
                weight.data.uniform_(-stdv, stdv)

    def initHidden(self, bsz):
        #Create hidden variable(s)
        raise NotImplementedError

    def resetHidden(self, bsz):
        #Re-wrap hidden variables
        raise NotImplementedError

    def initInference(self, bsz):
        #Re-wrap hidden in a variable with volatile=True
        raise NotImplementedError

    def forward(self, input):
        #Implement RNN layer and return output
        raise NotImplementedError
```

**Comment by csarofeen** [comment]
@apaszke @bmccann @adamlerer @SeanNaren If we can agree on an API I would be happy to provide some heavy lifting.

I think supporting backwards compatibility for direct RNN modules like https://github.com/adamlerer/examples/blob/onmt3/OpenNMT/onmt/Models.py#L19-L22 would be trivial, but continued support of https://github.com/pytorch/examples/blob/master/word_language_model/model.py#L9-L11 may simply be done for the sake of supporting that method for construction.

I think the function cells should go and each RNN type should be its own module. Autograd and Stacked functions could also be incorporated into a single module.

**Comment by jekbradbury** [comment]
It would be fairly difficult to make something like that work for the main `LSTM` and `RNN` classes while retaining full CUDNN compatibility. But the `LSTMCell` and `RNNCell` classes probably don't benefit performance-wise from using CUDNN, and could be made more transparent and easy to subclass (perhaps along the lines of Chainer, where the LSTM cell class is an ordinary module with two `Linear` submodules representing the "x to h" and "h to h" transformations). Then using the cell classes to build RNNs is not too difficult.

**Comment by csarofeen** [comment]
LSTMCell greatly benefits for me with cuDNN. Especially since it has fused pointwise operations and streamed GEMMs. Haven't directly benchmarked RNNCell.

Why would retaining cuDNN compatibility be difficult? Substitute the forward pass for a cuDNN based function if parameters are set to those that cuDNN can handle.

**Comment by apaszke** [comment]
@csarofeen LSTMCells never use cuDNN at the moment. You have to use the full `nn.LSTM` container for that.

I think this is a very important issue, and I wouldn't like to make any decisions in rush. I'd rather think for a while and come up with a very good API, that will be here to stay. Otherwise we can end up with a bunch of legacy stuff waiting to be deleted (or lots of broken checkpoints). If you guys have any ideas, code examples (like the one above, thanks a lot for that), wishes or comments, I think it's best to keep them in here or in a separate slack channel. We're probably going to discuss some of them this week.

**Comment by csarofeen** [comment]
Only comment about the above code is that I put the layers in a list in there which doesn't work with certain function calls, but since module's _modules is an ordered dict, seems it would be easy to add them in _modules and use them that way.

**Comment by hashbangCoder** [comment]
> Currently base RNN class/functions are hard to extend. If someone would like to extend LSTM with new features to pytorch they would have to modify:
AutogradRNN (nn/_functions/rnn.py)
StackedRNN (nn/_functions/rnn.py)
RNNBase (nn/modules/rnn.py)

 Could you explain this a bit? Likely I'm missing something, but my understanding was that it was trivial to modify LSTM cells as given [here]https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/rnn.py#L20). With that cell, shouldn't it be simple to loop through the sequence and execute the forward call each time and let autograd take care of the backward call? 

I'm planning to implement a simple variation the LSTM cell with a third input alongside input(t) and hidden(t-1).

@csarofeen @apaszke

**Comment by qyouurcs** [comment]
@hashbangCoder It's easy to modify LSTMCell class in order to implement new variants of LSTM. However, you cannot use the speedup from cuDNN as pointed by @apaszke, which means you will sacrifice some performance in terms of speed.

**Comment by matthew-z** [comment]
Any progress on this issue?

**Comment by ASalvail** [comment]
Or, on a related note, how can we implement gradient truncation from a certain time-step? 

It may not be super efficient, but I'm ready to unroll the graph with a for loop to at least get results until you have a better solution.

**Comment by guillaume-chevalier** [comment]
Does a clean and extendable LSTM implementation exists in PyTorch? 

https://stackoverflow.com/questions/50168224/does-a-clean-and-extendable-lstm-implementation-exists-in-pytorch

**Comment by zou3519** [comment]
@guillaume-chevalier Not yet. We're working on improving the implementation so that it is more extensible.

**Comment by icoxfog417** [comment]
From the aspect of class design, the dependency from the parent class to child classes is not appropriate. 
Specifically, [parent `RNNBase` class has the dependency on child classes `LSTM`, `GRU`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L40).

```py
        if mode == 'LSTM':
            gate_size = 4 * hidden_size
        elif mode == 'GRU':
            gate_size = 3 * hidden_size
        else:
            gate_size = hidden_size
```

If we want to extend new RNN cell (new `mode`), we have to modify parent class that will affect all the other child classes. And we have to select the `mode` when we make child class in spite of if it has no dependency on existing `mode`.

**Comment by guillaume-chevalier** [comment]
@icoxfog417 Yes, that's the point. And not only that, the fact there are 3 or more classes just to implement the RNN and RNN Cells functionnalities is a bit puzzling. 

In my jargon, requiring to add entries to such "if elif else" everywhere for new implementations is called poor polymorphism :stuck_out_tongue:. There is then the lack of new kind of classes to represent those "GRU" Cells, "LSTM" Cells, and other types of cells such as the default one. I'd suggest a base class named "RNNCell" or something like that. And then it would be possible to loop on those cells from the outside classes. There could also be some State classes. For example, see my implementation here of such diverse classes. See: 
- LARNN (RNN): https://github.com/guillaume-chevalier/Linear-Attention-Recurrent-Neural-Network/blob/df490bec6acc7949102daa38ca744a10a97a4ea8/larnn.py#L20
- LARNN Cell class: https://github.com/guillaume-chevalier/Linear-Attention-Recurrent-Neural-Network/blob/df490bec6acc7949102daa38ca744a10a97a4ea8/larnn.py#L150
- LARNN Cell State class: https://github.com/guillaume-chevalier/Linear-Attention-Recurrent-Neural-Network/blob/df490bec6acc7949102daa38ca744a10a97a4ea8/larnn.py#L108

Note: after reflexion on my own code, I'd rename the "LARNN" class (first in list above) to simply "RNN", such that it gets passed an instanciated cell class as argument to loop on. This would decouple the RNN concept from the LARNN Cell and the LARNN State and would make the RNN class suitable to receive any type of Cell and any type of State. You could probably also have a "StackedRNN" class in which we could inject an RNN class with its cell, something like that. Note that in my case here, the State is a queue (first in, last out) tensor on the previous states, on which I add a multi-head attention mechanism [see visualization](https://github.com/guillaume-chevalier/Linear-Attention-Recurrent-Neural-Network/blob/master/inkscape_drawings/png_exported/larnn-cell.png). 

It would therefore be interesting to have a "State" abstract pure class, because states are often quite different: the RNN has one inner state to pass, the LSTM has a tuple of inner states(cell state and hidden state), and my LARNN has a queue of cell states and one hidden state (like the LSTM, but the cell state is a queue). 

In the end, I needed to code this from scratch and loop myself over my own cell class, becaues I couldn't derive things from the PyTorch classes because of the poor polymorphism (and also the lack of clarity or proper OO architecture). However, it's not bad to derive such State classes and Cell classes and RNN base class, because from that people could write quite flexible and dynamic code. 

I hope this comment helps.

**Comment by danmackinlay** [comment]
Yes. I think that's more or less standard. AFAICT the messy structure in current Pytorch is, looking at code, about conveniently using the inconvenient and highly opinionated cuDNN optimisations, and it's hard to see how that would work with the nice clean inheritance structure above, which is much more general than cuDNN implements (I think it is at least - my eyes glaze when i read the cuDNN docs). Support ing both general user-defined RNNs and highly optimised pre-fab cuDNN optimisations and doing it all with nice clean inheritance and clear simple design is non-trivial. Which is, i presume, why we are where we are.

That said, if we give up the cudnn optimisations, it's precisely as simple as @guillaume-chevalier suggests, and I suspect that many of us have already implemented approximately that for our own code. I know I have, and it all fits in 2 short python modules. (Exception: I just use tuples rather than custom State classes, because I'm not enthusiastic about inheritance, but apart from that...)  Probably someone could cough up a decent implementation right now....

**Comment by moskomule** [comment]
Hi, any updates on this issue?
> We are working on letting people write `LSTMCell + for-loops` level RNN code that has CuDNN speeds
This sounds very cool.

**Comment by soumith** [comment]
it's not fast yet, and we're working on it, but this is the repo used for us benchmarking: https://github.com/pytorch/benchmark/tree/master/rnns/fastrnns

**Comment by apaszke** [comment]
Yes, packing functions work with the JIT just fine.

**Comment by matthew-z** [comment]
I was wondering why JIT could make looping cells as efficient as CUDNN implementations. 

If my understanding is correct, Pytorch JIT can reduce the python overhead in looping, and it is helpful only when the calculation in each iteration is very simple. I guess the overhead is quite small for some complex RNN structures (e.g., with attention mechanism). 

Also, even Tensorflow dynamic_rnn, which has no python overhead in graph mode, is much slower than CUDNN RNN implementation as the latter is fused over the time dimension.

**Comment by soumith** [comment]
@matthew-z there are several reasons we can achieve this, but will not elaborate right now. When we finish the work, we'll maybe write a  post about it.

**Comment by matthew-z** [comment]
@soumith Thanks!  I look forward to it.

**Comment by tangyuq** [comment]
I am trying to work out a minimal example of using customized RNNCell with pack_padded_sequence. Just want to make sure that the following pseudo code is correct. If not, can you please point to or create a jit example which works with packed APIs.

@torch.jit.script
def rnn_relu_cell(input, hidden, w_ih, w_hh, b_ih, b_hh):
    igates = torch.mm(input, w_ih.t()) + b_ih
    hgates = torch.mm(hidden, w_hh.t()) + b_hh
    return torch.relu(igates + hgates)

batch_cells = {}
@torch.jit.script
def rnn_relu(input, hidden, w_ih, w_hh, b_ih, b_hh):
     packed, batch_sizes = input
     seq_start = 0
     outputs = []
     output_batch_sizes = batch_sizes
     for batch_size in batch_sizes:
          cell = torch.jit.batch(cell, batch_size=batch_size) if batch_size not in batch_cells else batch_cells[batch_size]
          seq_end = start_start + batch_size 
          hidden[0:batch_size] = cell(packed[seq_start:seq_end], hidden[0:batch_size], w_ih, w_hh, b_ih, b_hh)   
          seq_start = seq_end  
          outputs.append(hidden[0:batch_size])
     return torch.stack(outputs), output_batch_sizes

rnn_relu = lambda input, hidden: rnn_relu(input, hidden, ....initiated parameters...)
batch_in = torch.zeros((batch_size, 1, max_length)) 
hidden = torch.zeros((batch_size, 1, 1)) 
pack = torch.nn.utils.rnn.pack_padded_sequence(batch_in, seq_lengths, batch_first=True)
out = rnn_relu(pack, hidden)
unpacked, unpacked_batch_sizes = torch.nn.utils.rnn.pad_packed_sequence(out)

**Comment by moskomule** [comment]
AFAIK, `hidden[0:batch_size] = cell(packed[seq_start:seq_end], hidden[0:batch_size], w_ih, w_hh, b_ih, b_hh)` is not supported in `jit` yet.

**Comment by Ricocotam** [comment]
Hi, any update on this ?

Does anyone have a link to extend RNNCells ?

If needed, I'd happy to help recoding what's been proposed

**Comment by soumith** [comment]
@Ricocotam see my comment here: https://github.com/pytorch/pytorch/issues/711#issuecomment-429072413

**Comment by voslucas** [comment]
Hi, i saw the comment. Is there already a (simple) example how to use it? I found this issue because i need a GRU (layer) with just a slightly different calculation in the internal cell. It's a minor change on paper, but it's hard to extend in practice... And yes, custom GRU/LSTM implementations can be found, but most of them lack the full RNNBase support (packed, dropout, layers, etc. )

**Comment by Ricocotam** [comment]
It's not clear if it uses packed_sequence (from what I saw by looking a few minutes, it doesn't). I think it's probably the most useful feature on custom RNNs

**Comment by soumith** [comment]
if you want variable-length RNNs, yes it uses packed-sequence. Otherwise, it doesn't need to use Packed sequences.

**Comment by ezyang** [comment]
I'm putting this into the jit triage queue because it seems like there is a case for some documentation / tutorials for writing JITable custom RNNs which doesn't seem to exist at the moment.

**Comment by daehwannam** [comment]
I made a LSTM frame, where custom LSTM is defined by a list of LSTM cell objects:
https://github.com/daehwannam/pytorch-rnn-util
https://stackoverflow.com/a/58658005/6710003

The custom LSTM supports all the __init__ options of PyTorch's standard LSTM but it must be slower than the standard LSTM.

**Comment by eellison** [comment]
Taking out of `jit` triage queue because we published a tutorial on JITable custom RNNs: https://pytorch.org/blog/optimizing-cuda-rnn-with-torchscript/
